✨ feat : 리뷰 생성 api 구현

Hyuk741 opened the pull request:
## ✅ PR 요약
- 관련 이슈 번호: #123
- 작업 요약: 리뷰생성 api 구현 및 테스트코드 작성 
- 테스트 시나리오
- 1)리뷰생성테스트
- 2)리뷰기본값테스트
- 3)중복방지테스트
- 4)여러사용자가 같은 리뷰 생성테스트
- 5)최대길이 테스트
- 6)review와 다른 모델과의 Fk관계확인 테스트

## 📄 상세 내용
- [시리얼라이저 ] reviews.py -> 리뷰 생성 api 구현
- [ 뷰]reviews.py-> 리뷰 생성 api 구현
- [ 테스트]reviews.py -> 리뷰생성에 관한 테스트 코드 작성

## 📸 스크린샷 (선택)

## 📝 기타 참고 사항

## 🧪 PR Checklist
- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

209512 committed:
✨ feat(lecture): bookmark serializers 추가

- 북마크 추가/삭제, 목록 조회를 위한 시리얼라이저

209512 committed:
🐛 fix(lecture): bookmark_serializers

base_crawler.py 예정에 따라 duration 함수 제거
LectureBookmarkButtonSerailizer 제거.
- 북마크 추가/삭제를 위한 별도의 Payload 없이 Path Parameter만 사용.
-> LectureBookmarkButtonSerializer는 제거하고 View에서 모델을 직접 조작.
- 추후 필요에 따라 다시 생성 후 유효성 검사 추가 가능.

209512 committed:
✨ feat(lecture): bookmark mock view, pagination mixin 추가

apps/core/utils/pagination_mixin.py:
- 공용 PaginationHandlerMixin 클래스 추가.

LectureBookmarkListAPIView에 PaginationHandlerMixin 적용.

LectureBookmarkButtonAPIView(북마크 추가/삭제): 불필요한 LectureBookmarkButtonSerializer 참조 및 사용 제거

209512 committed:
🎨 style(lecture): code formatting

209512 committed:
🐛 fix(lecture): bookmark mockdata api

bookmark views
- 북마크 목록 조회, 추가를 하나의 클래스(LectureBookmarkListCreateView)로 통합
- 북마크 삭제 API는 별도의 뷰(LectureBookmarkDeleteView)로 분리.
url 수정.
Serializers.py 수정.
- IntegrityError 예외 처리로 동시성 문제, 중복 생성 방지
- 북마크 생성 시 중복 체크를 위한 유효성 검사 로직(validate 메서드) 추가

rawonkim committed:
✨ feat: 인프런 카테고리 크롤러 구현

- BaseCrawler 타입 어노테이션 추가
- InflearnCategoryCrawler 구현
- 코드 품질 검사 통과

rawonkim committed:
Update apps/lecture/crawlers/base.py

💡 chore: 코드 리뷰 반영

Co-authored-by: 머용 <153697857+Meoyoug@users.noreply.github.com>

rawonkim committed:
Rename base.py to base_crawler.py

💡 chore: 코드 리뷰 반영

Commit:
feature/serializer보충

Commit:
fix/프론트요청에따른수정

rawonkim committed:
Update base_crawler.py

"♻️ refactor: 코드 리뷰 반영 
- API 전용 크롤러로 변경 
- 제이쓴 응답 데이터 무결성 검증 추가 "

rawonkim committed:
Update base_crawler.py

🎨 style: import 정렬 수정

Hyuk741 committed:
✨ feat : 리뷰 생성 api 구현

Hyuk741 committed:
✨ feat : 리뷰 생성 api 구현

209512 committed:
✅ test(lecture): 강의 북마크 통합 테스트 코드 추가

# Conflicts:
#	apps/lecture/tests/__init__.py

209512 committed:
🎨 style(lecture): code formatting 수정

choismjames23 commented:
리베이스를 잘못 진행하신거 같아요! 커밋이 기존거랑 합쳐져서 123개가 올라왔습니다. 확인 부탁드려요!

Hyuk741 committed:
Merge branch 'develop' into feature/reviews_api

Hyuk741 committed:
✨ feat : 리뷰 생성 api 구현

# Conflicts:
#	apps/studies/urls.py

Hyuk741 committed:
✨ feat : 리뷰 생성 api 구현

# Conflicts:
#	apps/studies/urls.py

Hyuk741 committed:
✨ feat : 리뷰 생성 api 구현

rawonkim committed:
💡 chore: 파일명 변경에따른 import 경로 수정 (base → base_crawler)

Meoyoug committed:
🚚 build: 패키지 인식을 위해 __init__.py 추가

Meoyoug committed:
💡 chore: test 코드 파일 이름 수정

Meoyoug committed:
🚚 build: tblib 추가

209512 committed:
🐛 fix(lecture): 북마크 생성 시 serializer.save(user=request.user) 방식 적용

- 북마크 생성 시 user 정보를 시리얼라이저 save() 메서드 인자로 명시적으로 전달하도록 변경
- serializer에서 context["request"].user 참조 제거
- get_or_create()로 중복 생성 처리
- 뷰 POST 핸들러에서 context 전달 제거, 예외 처리 변경

Meoyoug committed:
💡 chore: unused  "type: ignore" comment delete.

Meoyoug committed:
✅ test: IsolatedCacheTestClient 를 사용하도록 변경

Commit:
fix/n+1수정

Meoyoug committed:
🚚 build: install django-filter-stubs v0.1.3

Meoyoug committed:
✅ test: 사용되지 않는 테스트 코드 파일 삭제

Meoyoug committed:
✨ feat: __init__.py 추가

Meoyoug commented on apps/studies/tests/reviews.py:
test 코드 파일명은 test_~~.py 형태로 맞춰주셔야합니다. 그래야 test 명령어 실행시 해당 파일을 읽어와서 테스트코드를 실행해요.

추가로 해당 테스트코드 내용은 Model Test 성격에 가까운 것 같습니다.

test client 를 사용하여 API 호출을 한 다음, 실제 응답이 정상적으로 오는 지, 적절한 예외가 발생하고 예외응답이 오는 지 등을 테스트 하셔야합니다.

Meoyoug commented on apps/studies/urls.py:
```diff
@@ -1,12 +1,17 @@
 
 urlpatterns = [
+    # POST /api/v1/studies/groups/{group_id}/reviews/
+    path("studies/groups/<uuid:group_id>/reviews/", ReviewCreateView.as_view(), name="group-review-create"),
```
studies와 groups가 다른 데이터 항목을 가리키나요..?
같다면 굳이 나누지않아도 되어 보이는데요

Meoyoug commented on apps/studies/serializers/reviews.py:
```diff
@@ -0,0 +1,73 @@
+        user_id: str | int = raw_pk if isinstance(raw_pk, (str, int)) else str(raw_pk)
+        study_group = attrs["study_group"]
+
+        if Review.objects.filter(user_id=user_id, study_group=study_group).exists():
```
이부분은 View나 서비스 로직에서 처리되어야하는 분기인것같습니다.

Meoyoug commented on apps/studies/serializers/reviews.py:
```diff
@@ -0,0 +1,73 @@
+
+        raw_pk = getattr(request.user, "pk", None)
+        if raw_pk is None:
+            raise serializers.ValidationError({"non_field_errors": ["인증이 필요합니다"]})
```
이부분은 View의 permission_classes 에 의해서 처리되는 부분입니다.

fullrestic committed:
🐛 fix: 병합 충돌 해결

209512 committed:
🐛 fix(lecture): 북마크 통합 테스트의 Mocking 제거 및 로직 일치화

- Mocking 사용 전면 제거: 테스트 코드에서 `patch`를 이용한 Mocking 사용을 모두 제거. 실제 뷰 로직 및 DB 동작에 기반한 통합 테스트로 전환. (외부 API 호출이 아닌 자가 작성 코드에 대한 Mocking 부적절 피드백 반영)

- POST IntegrityError 검증: `serializer.save()` 시 중복 북마크 생성으로 인해 발생하는 `IntegrityError`가 DRF에 의해 400 에러로 정확히 변환/처리되는지 실제 DB 조작을 통해 검증하도록 로직 수정.

- GET 페이지네이션 검증 강화: 뷰 파일의 Mock 데이터(총 15개)를 기준으로 1페이지(10개)뿐만 아니라 2페이지(5개)의 데이터 개수 및 중복 여부까지 명확히 확인하도록 테스트 보완.

- POST 에러 처리 로직 일치화: `serializer.save()` 중 발생하는 예외 처리 시, 뷰의 로직에 맞춰 테스트 기대값 수정.

Commit:
feature/SpecAPI검증test추가

fullrestic committed:
♻️ refactor : 휴대폰 인증 API 목적별 분리

fullrestic committed:
💡 chore    : User Manager create_user 임의 추가

fullrestic committed:
💡 chore    : 키 생성시 정규화 번호 사용하도록 수정

fullrestic committed:
♻️ refactor : auth_urls로 인증 Url 병합

Meoyoug committed:
Test IsolatedCacheTestClient Update

Meoyoug committed:
change test code

Meoyoug committed:
🚚 build: 장고 Test 명령어 오버라이딩, CI 수정

Meoyoug committed:
add type annotations, run black, isort

Meoyoug committed:
fix: phone verification endpoint 변경

Hyuk741 committed:
🐛 fix : 리뷰 생성API 피드백 반영 수정본
